src/main/java/io/pravega/perf/PerfStats.java

```diff
@@ -94,138 +103,245 @@ public void print(Instant time) {
+                        .withFirstRecordAsHeader().withIgnoreHeaderCase().withTrim());
+
+                for (CSVRecord csvEntry : csvParser) {
+                    record(Integer.parseInt(csvEntry.get(1)), Integer.parseInt(csvEntry.get(2)));
+                }
+                csvParser.close();
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
 
-        if (this.count % this.sampling == 0) {
-            this.latencies[index] = latency;
-            this.index++;
+        @Override
+        public void record(long start, int bytes, int latency) {
+            try {
+                csvPrinter.printRecord(start, bytes, latency);
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
+            }
         }
 
-        if (this.printer != null) {
-            final long nanotime = startTime.getEpochSecond() * NANOSEC_PER_SEC + startTime.getNano();
-            printer.printRecord(count, event, bytes, nanotime, latency);
+        @Override
+        public void printTotal(long endTime) {
+            try {
+                csvPrinter.close();
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
+            }
+            readCSV();
+            super.printTotal(endTime);
         }
     }
 
-    private static long[] percentiles(long[] latencies, int count, double... percentiles) {
-        int size = Math.min(count, latencies.length);
-        Arrays.sort(latencies, 0, size);
-        long[] values = new long[percentiles.length];
-        for (int i = 0; i < percentiles.length; i++) {
-            int index = (int) (percentiles[i] * size);
-            values[i] = latencies[index];
+    public PerfStats(String action, int reportingInterval, int messageSize, String csvFile) {
+        this.action = action;
+        this.messageSize = messageSize;
+        this.windowInterval = reportingInterval;
+        this.csvFile = csvFile;
+        this.queue = new ConcurrentLinkedQueue<TimeStamp>();
+        this.executor = new ForkJoinPool(1);
+    }
+
+    /**
+     * private class for start and end time.
+     */
+    private class ProcessQueue implements Runnable {
+        public void run() {
+            TimeStamp t;
+            while (true) {
+                try {
+                    t = queue.poll();
+                    if (t != null) {
+                        record(t.start, t.end, t.bytes);
+                    } else {
+                        LockSupport.parkNanos(100);
+                    }
+                    print();
+                } catch (IOException ex) {
+                    throw new RuntimeException(ex);
+                }
+            }
         }
-        return values;
+    }
+
+
+    private void record(long startTime, long endTime, int bytes) {
+        final int latency = (int) (endTime - startTime);
+        window.record(bytes, latency);
```

Comment:
This occurs the the executor thread with no lock held. So an NPE could be thrown right here.

Reply:
this record is always called by dedicated thread;  multiple threads wont access this method.